proxy.go

```diff
@@ -188,6 +216,207 @@ func (proxy *proxy) releaseToken(token Token) (*tokenInfo, error) {
 	return info, nil
 }
 
+var storeStateDir = "/var/lib/clear-containers/proxy/"
+
+type proxyStateOnDisk struct {
+	SocketPath      string   `json:"socket_path"`
+	EnableVMConsole bool     `json:"enable_vm_console"`
+	ContainerIDs    []string `json:"container_ids"`
+}
+
+// returns false if it's a clean start (i.e. no state is stored) or restoring failed
+func (proxy *proxy) restoreState() bool {
+	proxyStateFilePath := storeStateDir + "proxy_state.json"
+	if _, err := os.Stat(storeStateDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(storeStateDir, 0755); err != nil {
+			proxyLog.Errorf("Couldn't create directory %s: %v",
+				storeStateDir, err)
+		}
+		return false
+	}
+	if _, err := os.Stat(proxyStateFilePath); os.IsNotExist(err) {
+		return false
+	}
+
+	fdata, err := ioutil.ReadFile(proxyStateFilePath)
+	if err != nil {
+		proxyLog.Errorf("Couldn't recover from %s: %v", proxyStateFilePath, err)
+		return false
+	}
+
+	var proxyState proxyStateOnDisk
+	err = json.Unmarshal(fdata, &proxyState)
+	if err != nil {
+		proxyLog.Errorf("Couldn't unmarshal %s: %v", proxyStateFilePath, err)
+		return false
+	}
+	proxyLog.Debugf("proxy: %+v", proxyState)
+
+	if len(proxyState.ContainerIDs) == 0 {
+		return false
+	}
+	proxyLog.Warn("Recovering proxy state from: ", proxyStateFilePath)
+	proxy.socketPath = proxyState.SocketPath
+	proxy.enableVMConsole = proxyState.EnableVMConsole
+
+	for _, contID := range proxyState.ContainerIDs {
+		go restoreVMState(proxy, contID)
+	}
+
+	return true
+}
+
+func (proxy *proxy) storeState() {
+	proxyStateFilePath := storeStateDir + "proxy_state.json"
+	proxy.Lock()
+	defer proxy.Unlock()
+
+	// if there are 0 VMs then remove state from disk
+	if (len(proxy.vms)) == 0 {
+		if err := os.Remove(proxyStateFilePath); err != nil {
+			proxyLog.Errorf("Can not remove %s: %v", proxyStateFilePath, err)
+		}
+		return
+	}
+
+	proxyState := &proxyStateOnDisk{
+		SocketPath:      proxy.socketPath,
+		EnableVMConsole: proxy.enableVMConsole,
+		ContainerIDs:    make([]string, 0, len(proxy.vms)),
+	}
+	for cid := range proxy.vms {
+		proxyState.ContainerIDs = append(proxyState.ContainerIDs, cid)
+	}
+
+	data, err := json.MarshalIndent(proxyState, "", "\t")
+	if err != nil {
+		proxyLog.Errorf("Couldn't marshal proxy state %+v", proxyState)
+	}
+	if err = ioutil.WriteFile(proxyStateFilePath, data, 0644); err != nil {
+		proxyLog.Errorf("Couldn't store proxy state to %s: %v",
+			proxyStateFilePath, err)
+	}
+}
+
+// Represents vm struct on disk
+type vmStateOnDisk struct {
```

Comment:
Again, this would be better "up-front" at the top of the file.

Reply:
Will do.